### ISSUE TITLE
18-07-2017: Inheritance inconsistency

### ISSUE THREAD

**Issue by vitkarpov** [jakwuh/dailytip#2]
Am I right in assuming that what you've described is an error and should be fixed? If so, did you make a pr or something to Webpack?

**Comment by jakwuh** [comment]
tldr; the described problem comes from historical JS inconsistency, not webpack or babel bug;

When you construct an object by calling `new` on a function, you have two possible ways of instantiation:

```js
function Test() {
  this.a = 2;
}

new Test() instanceof Test; // true

function Test2() {
  return {a: 2}
}

new Test2() instanceof Test2; // false
```

This is not the case for classes - which does not support returning anything from `constructor`. So, `new ClassName() instanceof ClassName` will **always** return `true`.

As you might guess, a native `Error` class returns the `object` during instantiation. This makes the `Error` function incompatible with classes syntax. What babel does in this case - it simply uses the constructed by `Error` value as `this` reference, which means `instanceof` will only return `true` when called on `Error` class. This is not a bug but a decision about dealing with JS inconsistency.

**Comment by vitkarpov** [comment]
Gotcha! ;)

> This is not a bug but a decision about dealing with JS inconsistency.

Wait a minute, why I can do this with regular JS like this:

```js
function CommonError() {
  this.constructor.name = CommonError;
}
CommonError.prototype = Object.create(Error.prototype);

function HTTPError() {
  this.constructor.name = HTTPError;
}
HTTPError.prototype = Object.create(CommonError.prototype);

let error = new HTTPError();
// true
console.log(error instanceof Error);
// true
console.log(error instanceof CommonError);
// true
console.log(error instanceof HTTPError);
```

but can't do with Webpack which is dealing with classes underneath? I guess I'm missing something.

**Comment by vitkarpov** [comment]
And to be sure we're on the same page, what do you think would be the right behavior? (I mean by design, doesn't matter how it works and doesn't right now)

**Comment by jakwuh** [comment]
@vitkarpov in your example you miss a `super` constructor (or function) call (`HTTPError` should call `CommonError` function, which in turn should call `Error` function). Once you try to fix it, you will need to decide how to deal with functions which return `objects` (e.g. `Error`).

Without considering JS history, I'd like JS to ignore function return value when it's called with `new` operator.

**Comment by vitkarpov** [comment]
@jakwuh yeah, get it now, bummer!

```js
function CommonError() {
  this.constructor.name = CommonError;
  return Error.call(this);
}
CommonError.prototype = Object.create(Error.prototype);

function HTTPError() {
  this.constructor.name = HTTPError;
  return CommonError.call(this);
}
HTTPError.prototype = Object.create(CommonError.prototype);

let error = new HTTPError();
// true
console.log(error instanceof Error);
// false
console.log(error instanceof CommonError);
// false
console.log(error instanceof HTTPError);
```

Thanks for your explanation :)
